src/query.rs

```diff
@@ -69,6 +77,46 @@ use errors::{
 };
 
 pub type QueryExecutionResult = Result<QueryResults>;
+pub type PreparedResult<'conn, 'sqlite> = Result<PreparedQuery<'conn, 'sqlite>>;
+
+pub enum PreparedQuery<'conn, 'sqlite> {
+    Empty {
+        conn: &'conn Conn,
+        schema: Schema,
+        find_spec: FindSpec,
+    },
+    Bound {
+        conn: &'conn Conn,
+        schema: Schema,
+        statement: rusqlite::Statement<'sqlite>,
+        args: Vec<(String, Rc<rusqlite::types::Value>)>,
+        projector: Box<Projector>,
+    },
+}
+
+impl<'conn, 'sqlite> PreparedQuery<'conn, 'sqlite> {
+    pub fn run<T>(&mut self, _inputs: T) -> QueryExecutionResult where T: Into<Option<QueryInputs>> {
+        match self {
+            &mut PreparedQuery::Empty { ref conn, ref schema, ref find_spec } => {
+                let current_schema = &*conn.current_schema();
```

Comment:
I don't believe you need the schema checks right now.

At present you're scoping a `PreparedQuery` to an `InProgress(Read)`, which is isolated.

In the future `PreparedQuery` (or a related kind of prepared query!) instances might be scoped to a `Conn`, in which case they can be invalidated when the schema changes… but I think we can reduce the coupling required by that check. (See discussion below.)

Your choice whether you want to do the narrow-scoped version, the broad-scoped version, or both.

Reply:
I thought about doing the broad-scoped version, but it's tricky to scope a `PreparedQuery` to a `Conn`, I think because `Conn` doesn't currently manage the `rusqlite::Connection`. I tried this out with a lifetime annotation like `PreparedQuery<'conn, 'sqlite>`, but that confuses the borrow checker: either the `conn` or the `statement` doesn't live long enough.

Reply:
Remember we discussed coupling the metadata/conn and the SQLite connection?

Right now `InProgress` and `InProgressRead` are a pairing of SQLite and metadata, with the side-effect of also managing a transaction. That is: they manage a `rusqlite::Transaction`, not a `rusqlite::Connection`.

I think there's another concept here, which is the pairing of a `rusqlite::Connection` and metadata. That might, as you say, be `Conn`, or a sibling of it.

I think the narrow-scoped version here is fine for now; we can generalize later.